### PR TITLE
Stub out pthread_rwlock_trywrlock and pthread_rwlock_tryrdlock

### PR DESCRIPTION
**Pull request by Aaron1011**

### CODE REVIEW

**Comment by oli-obk** [comment]
do you have any code using just the libstd that will excercise these functions?

**Comment by RalfJung** [comment]
The `try_read`/`try_write` methods of `RwLock` call this, so it should be possible to extend https://github.com/rust-lang/miri/blob/master/tests/run-pass/sync.rs to test this.

However, see https://github.com/rust-lang/miri/issues/781: the existing shims, and the new ones, are not actually correct, not even for sequential code. With the existing shims, only code that would *always* diverge into an endless loop or panic is affected. However, with the new shims, it is trivial to write some reasonable sequential code that will go wrong:
```rust
use std::sync::RwLock;

fn main() {
    let l = RwLock::new(42);
    let g = l.read();
    assert!(l.try_write().is_err());
}
```

In other words, so far the bug only affects code that is rather silly, but this PR even reasonable code can run into this bug. So I am not very comfortable landing this as-is.

**Comment by RalfJung** [comment]
So @Aaron1011 maybe you could help fix https://github.com/rust-lang/miri/issues/781 ? My thinking is that we do have some amount of state available there (namely, one pointer-size worth of state) and we can just use that to encode the state of the Mutex/RwLock. We'll have to be careful about alignment, which costs is 3 bits of state or so, but there's still plenty left. It's kind of crude, but this way we could avoid having to do an extra allocation for this.

We could of course also do an extra allocation, that just feels like more work to me.

**Comment by RalfJung** [comment]
Closing for now due to lack of feedback; @Aaron1011 feel free to reopen if you want to pick this up again.
